torch/quantization/_numeric_suite.py

```diff
@@ -39,3 +42,139 @@ def compare_weights(float_dict, quantized_dict):
+        super(Logger, self).__init__()
+        self.stats = {}
+        self.stats["float"] = None
+        self.stats["quantized"] = None
+
+    def forward(self, x, y):
+        if self.stats["float"] is None:
+            if x.is_quantized:
+                self.stats["quantized"] = x.dequantize().detach()
+            else:  # Output is in float for dynamic quantization
+                self.stats["quantized"] = x.detach()
+
+            self.stats["float"] = y.detach()
+        else:
+            if x.is_quantized:
+                self.stats["quantized"] = torch.cat(
+                    (self.stats["quantized"], x.dequantize().detach())
+                )
+            else:  # Output is in float for dynamic quantization
+                self.stats["quantized"] = torch.cat(
+                    (self.stats["quantized"], x.detach())
+                )
+
+            self.stats["float"] = torch.cat((self.stats["float"], y.detach()))
+
+
+class Shadow(nn.Module):
+    r"""Shadow module attaches the float module to its matching quantized module
+    as the shadow. Then it uses Logger module to process the outputs of both
+    modules to do the comparison.
+    """
+
+    def __init__(self, q_module, float_module, Logger):
+        super(Shadow, self).__init__()
+        self.orig_module = q_module
+        self.shadow_module = float_module
+        self.dequant = nnq.DeQuantize()
+        self.logger = Logger()
+
+    def forward(self, x):
+        output = self.orig_module(x)
+        x = x.dequantize()
+        shadow_output = self.shadow_module(x)
+        self.logger(output, shadow_output)
+        return output
+
+
+def _prepare_model_with_stubs(float_module, q_module, module_swap_list, Logger):
+    r"""Prepare the model by attaching the float module to its matching quantized
+    module as the shadow if the float module type is in module_swap_list.
+
+    Args:
+        float_module: the float module used to generate the q_module
+        q_module: the quantized module
+        module_swap_list: list of float module types to attach the shadow
+        Logger: the class to be used in shadow module to process the outputs of
+            quantized module and its float shadow module
+    """
+
+    float_module_children = {}
+    for name, mod in float_module.named_children():
+        float_module_children[name] = mod
+
+    reassign = {}
+    for name, mod in q_module.named_children():
+        if name not in float_module_children:
+            continue
+
+        float_mod = float_module_children[name]
+
+        if type(float_mod) not in module_swap_list:
+            _prepare_model_with_stubs(float_mod, mod, module_swap_list, Logger)
+
+        if type(float_mod) in module_swap_list:
+            reassign[name] = Shadow(mod, float_mod, Logger)
+
+    for key, value in reassign.items():
+        q_module._modules[key] = value
+
+
+def compare_model_stub(float_model, q_model, module_swap_list, data, Logger=Logger):
```

Comment:
Wondering if we need this as a top level function or keep this as an internal/test only function and move it to the test code.

Reply:
Maybe we can keep it as top level function since it's easy for user to use as a one-liner to quickly check the model accuracy?

Reply:
Sure, as long as we have the building blocks as user facing functions, we are ok.